states/new_hampshire/main.py

```diff
@@ -0,0 +1,77 @@
+import json
+import scrapy
+from scrapy.crawler import CrawlerProcess
+import csv
+import re
+from io import StringIO
+from common import get_clerk_data_default_entry
+
+# We use Scrapy for New Hampshire because it has an .aspx __viewstate 
+# which must be generated and sent along with the request
+
+class NewHampshireSpider(scrapy.Spider):
+  name = 'New Hampshire'
+  start_urls = ["https://app.sos.nh.gov/Public/Reports.aspx"]
+
+  def parse(self, response):
+    yield scrapy.FormRequest.from_response(
+      response,
+      formdata={
+          'ctl00$MainContentPlaceHolder$PPReport': 'rdoCsv'
+      },
+      method="POST",
+      callback=self.generate_csv
+    )
+
+  def generate_csv(self, response):
+    f = StringIO(response.text)
+    reader = csv.DictReader(f, delimiter=",")
+    clerk_data = []
+
+    for row in reader:
+      data = self.extract_clerk_data(row)
+      if data is not None:
+        clerk_data.append(data)
+
+    with open('../../public/new_hampshire.json', 'w') as f:
+        json.dump(clerk_data, f)
+
+    return clerk_data
+
+  def extract_clerk_data(self, row):
+    town_or_city = self.extract_city_without_ward(row)
+
+    if town_or_city is None:
+      return None
+
+    clerk_data_entry = get_clerk_data_default_entry()
+    clerk_data_entry["city"] = town_or_city
+    clerk_data_entry["address"] = row["Address"].strip() 
+    clerk_data_entry["locale"] = town_or_city
+    email_pattern = re.compile(r"^([a-zA-Z0-9_\-\.]+)@([a-zA-Z0-9_\-\.]+)\.([a-zA-Z]{2,5})$")
+
+    if(email_pattern.match(row["E-Mail"])):
+      clerk_data_entry["emails"] = [row["E-Mail"]]
+    if(row["Phone (area code 603)"]):
+      clerk_data_entry["phones"] = ["603-" + row["Phone (area code 603)"]]
+    if(row["Fax"]):
+      clerk_data_entry["faxes"] = ["603-" + row["Fax"]]
+    if row["Clerk"]:
+      clerk_data_entry["official"] = row["Clerk"]
+
+    return clerk_data_entry
+
+  def extract_city_without_ward(self, row):
+    capitalized_city = " ".join(x.lower().capitalize() for x in row["Town/City"].split(" "))
+    city_split_by_ward = capitalized_city.split(" Ward ")
```

Comment:
Consider matching on regex `' Ward ' in x`.  If it matches, throw away row.  Otherwise, keep entire string (with your normalization above).

Reply:
we still need to extract data from the first ward, which contains ' Ward '. I cleaned up the logic to make this clearer

Reply:
Good point!  Thanks for catching that.